sling/nlp/embedding/fact-plausibility.cc

```diff
@@ -0,0 +1,547 @@
+      if (groups.length() >= min_facts_) {
+        // Split into training and evaluation set.
+        if (eval_ratio_ == 0 || rnd.UniformProb() > eval_ratio_) {
+          training_instances_.push_back(instance.handle());
+          num_train_instances->Increment();
+        } else {
+          evaluation_instances_.push_back(instance.handle());
+          num_eval_instances->Increment();
+        }
+      } else {
+        num_skipped_instances->Increment();
+      }
+      num_facts->Increment(facts.length());
+      num_fact_groups->Increment(groups.length());
+
+      delete message;
+    }
+    store_.Freeze();
+
+    // Run training.
+    LOG(INFO) << "Starting training";
+    Train(task, &model);
+
+    // Output profile.
+    LogProfile(model);
+
+    // Save final model.
+    if (!model_filename_.empty()) {
+      LOG(INFO) << "Saving model to " << model_filename_;
+      Save(model, model_filename_);
+    }
+
+    delete optimizer_;
+  }
+
+  // Worker thread for training embedding model.
+  void Worker(int index, Network *model) override {
+    // Initialize random number generator.
+    Random rnd;
+    rnd.seed(seed_ + index);
+
+    // Premises and hypotheses for one batch.
+    std::vector<Features> premises(batch_size_);
+    std::vector<Features> hypotheses(batch_size_);
+
+    // Set up plausibility scorer.
+    Instance scorer(flow_.scorer);
+    Instance gscorer(flow_.gscorer);
+    int *premise = scorer.Get<int>(flow_.premise);
+    int *hypothesis = scorer.Get<int>(flow_.hypothesis);
+    float *logits = scorer.Get<float>(flow_.logits);
+    float *dlogits = gscorer.Get<float>(flow_.d_logits);
+    std::vector<Instance *> gradients{&gscorer};
+
+    for (;;) {
+      // Compute gradients for epoch.
+      gscorer.Clear();
+      gscorer.Set(flow_.primal, &scorer);
+      float epoch_loss = 0.0;
+      Benchmark batch_benchmark;
+
+      for (int b = 0; b < batches_per_update_; ++b) {
+        // Random sample instances for batch.
+        for (int i = 0; i < batch_size_; ++i) {
+          int sample = rnd.UniformInt(training_instances_.size());
+          Frame instance(&store_, training_instances_[sample]);
+
+          // Split instance into premise and hypothesis.
+          int num_groups = instance.Get(p_groups_).AsArray().length();
+          int heldout = rnd.UniformInt(num_groups);
+          Split(instance, &premises[i], &hypotheses[i], heldout);
+        }
+
+        // Do forward and backward propagation for each premise/hypothesis pair.
+        // Each sampled item is a positive example. Negative examples are
+        // generated by using the premise from one item and the hypothsis from
+        // another item.
+        for (int i = 0; i < batch_size_; ++i) {
+          for (int j = 0; j < batch_size_; ++j) {
+            // Check that main hypothesis is not in premise for negative
+            // examples.
+            if (i != j && Has(premises[i], hypotheses[j][0])) {
```

Comment:
Is it correct to say that 'main hypothesis', i.e. the one at index 0, is the most fine-grained (i.e. not expanded) version of the fact?

Reply:
Yes, the base fact is the first in the group.